dandi/dandiapi.py

```diff
@@ -301,25 +305,78 @@ def paginate(
         ``"next"`` key until it is ``null``.
+
+        If the first ``"next"`` key is the same as the initially-requested URL
+        but with the ``page`` query parameter set to ``2``, then the remaining
+        pages are fetched concurrently in separate threads, ``lookahead``
+        (default 5) at a time.  This behavior requires the initial response to
+        contain a ``"count"`` key giving the number of items across all pages.
         """
         if page_size is not None:
             if params is None:
                 params = {}
             params["page_size"] = page_size
-        r = self.get(path, params=params, **kwargs)
-        while True:
-            for item in r["results"]:
-                yield item
-            if r.get("next"):
-                r = self.get(r["next"])
-            else:
-                break
+
+        resp = self.get(path, params=params, json_resp=False)
+        r = resp.json()
+        if r["next"] is not None:
+            page1 = resp.history[0].url if resp.history else resp.url
+            if not is_page2_url(page1, r["next"]):
+                lgr.warning(
+                    "Pagination request to %s returned unexpected 'next' URL: %s",
+                    page1,
+                    r["next"],
+                )
+                if os.environ.get("DANDI_PAGINATION_DISABLE_FALLBACK"):
+                    raise RuntimeError(
+                        f"API server changed pagination strategy: {page1} URL"
+                        f" is now followed by {r['next']}"
+                    )
+                else:
+                    while True:
+                        yield from r["results"]
+                        if r.get("next"):
+                            r = self.get(r["next"])
+                        else:
+                            return
+        yield from r["results"]
+        if r["next"] is None:
+            return
+
+        if page_size is None:
+            page_size = len(r["results"])
+        pages = (r["count"] + page_size - 1) // page_size
+        pageno_iter = iter(range(2, pages + 1))
+
+        with ThreadPoolExecutor(max_workers=lookahead) as pool:
+            futures: deque[Future[list]] = deque()
+
+            def get_page(pageno: int) -> list:
+                params2 = params.copy() if params is not None else {}
+                params2["page"] = pageno
+                results = self.get(path, params=params2)["results"]
+                assert isinstance(results, list)
+                return results
+
+            def submit_job() -> None:
+                try:
+                    i = next(pageno_iter)
+                except StopIteration:
+                    return
+                futures.append(pool.submit(get_page, i))
+
+            for _ in range(lookahead):
+                submit_job()
+            while futures:
+                res = futures.popleft().result()
+                submit_job()
+                yield from res
```

Comment:
why not just prepare futures for all pages ahead of time? i.e. smth like
```suggestion
            for i in range(2, pages + 1):        
                futures.append(pool.submit(get_page, i))

            while futures:
                yield futures.popleft().result()
```

Reply:
Because, if the user stops iterating partway through the results, your way would lead to every page still being fetched, whereas the current code would only fetch five more pages.

Reply:
hm, why would it? (self-educational question) I would have expected that leaving ThreadPoolExecutor disables execution of anything, so no other one would be operating. But if needed really -- should just cancel all futures I guess. I did similar in https://github.com/datalad/datalad/blob/master/datalad/support/parallel.py#L226 and that should then not leave any running... so I guess smth like

```python
            try:
                for i in range(2, pages + 1):        
                    futures.append(pool.submit(get_page, i))
                while futures:
                    yield futures.popleft().result()
            finally:
                while futures:
                    futures.popleft().cancel()
```
so even those five do not survive or in other words -- worth doing for your version as well... but I still wonder why/if that is needed... will try now

Reply:
FWIW -- seems similar pattern to e.g. (besides that there it is a "simpler" `.as_completed`) (also with cancels) as https://github.com/dandi/dandi-cli/blob/HEAD/dandi/files.py#L1218 ...

Reply:
The pool's `__exit__` method is only called when control reaches the end of the `with ThreadPoolExecutor:` block.  If the caller abandons the pagination partway through, execution of the `pagination()` generator will still be in the `while futures:` loop, and the jobs in the pool won't be stopped, and they'll continue to use up resources until garbage collection (or reference counting or whatever) does its thing. 

Similarly, with your suggested code, if the caller abandons the pagination partway through, the `while futures: yield futures.popleft().result()` will never finish, and so the `finally:` won't be executed.  The only way to ensure that the thread pool is shutdown when iteration is over is to call the paginator like:

```
from contextlib import closing

with closing(client.paginate(...)) as it:
    for obj in it:
         ...
```

and then every iterator that "yields from" the paginator (like `get_assets()`) would also have to be wrapped in `closing()` as well, along with any iterators that wrapped those, etc.

Reply:
<details>
<summary>confirmed with this script</summary> 

```python
from collections import deque
from concurrent.futures import Future, ThreadPoolExecutor

def get_page(page: int):
    print(f"getting page {page}")
    import time
    time.sleep(0.5)
    print(f"got page {page}")
    return f"returning page {page}"

lookahead = 5
pages = 100

with ThreadPoolExecutor(max_workers=lookahead) as pool:
    futures: deque[Future[list]] = deque()
    try:
        for i in range(2, pages + 1):
            futures.append(pool.submit(get_page, i))
        print(f"Created all {len(futures)}")
        while futures:
            print(f"Got {futures.popleft().result()}")
    finally:
        if True:
            print(f"CANCELING remaining {len(futures)}")
            while futures:
                futures.popleft().cancel()

print("DONE")


```
</details>
that canceling is indeed needed and that otherwise (`if False` in finally) it keeps plowing... that is odd since makes little sense to me when would I want such behavior while using the `ThreadPoolExecutor`

Reply:
OK, I just tried running this as a demonstration:

```python
import sys

def gen():
    try:
        for i in range(5):
            yield i
    finally:
        print(sys.exc_info())
        print("This will never be printed.")

for x in gen():
    print(x)
    if x == 3:
        break

"""
# Alternative:
it = gen()
print(next(it))
print(next(it))
print(next(it))
print("After")
"""
```

And Python ended up automatically closing the generator (specifically, raising `GeneratorExit` within it) after the `break` (and, for the "alternative" block, after "After" was printed).  I'm going to have to look into exactly under what circumstances that happens.

Reply:
@yarikoptic So then are you fine with the code as it is now?

Reply:
I am not yet catching exactly how it (investigation of "incompletely" explored generator) would relate besides showing that `finally` would always be executed (desired).

FWIW: in `finally` there could simply be `pool.shutdown(cancel_futures=True)` which would cancel the futures.

Reply:
> @yarikoptic So then are you fine with the code as it is now?

I would prefer a more consistent with other similar codes pattern of `try`/`finally: pool.shutdown(cancel_futures=True)`.

Reply:
Done (using `f.cancel()`, since `cancel_futures` was only added in Python 3.9).